src/concurrency.md

```diff
@@ -0,0 +1,760 @@
+
+The code is very similar to the one that disabled all interrupts except for the
+start and end of the critical section, which now include a `compiler_fence`
+(compiler barrier).
+
+### Priorities
+
+Architectures like ARM Cortex-M allow interrupt prioritization, meaning that an
+interrupt that's given high priority can preempt a lower priority interrupt
+handler. Priorities must be considered when sharing state between interrupt
+handlers.
+
+When two interrupt handlers, say `A` and `B`, have the *same* priority no
+preemption can occur. Meaning that when signals for both interrupts arrive
+around the same time then the handlers will be executed sequentially: that is
+first `A` and then `B`, or vice versa. In this scenario, both handlers can
+access the same `static mut` variable *without* using a critical section; each
+handler will "take turns" at getting exclusive access (`&mut-`) to the static
+variable. Example below.
+
+``` rust
+{{#include ../ci/concurrency/examples/coop.rs}}```
+
+When two interrupt handlers have *different* priorities then one can preempt
+the other. Safely sharing state between these two interrupts requires a critical
+section in the lower priority handler -- just like in the case of `main` and an
+interrupt handler. However, one more constraint is required: the priority of the
+interrupts must remain fixed at runtime; reversing the priorities at runtime,
+for example, would result in a data race.
+
+The following example showcases safe state sharing between two interrupt
+handlers using a priority-based critical section.
+
+``` rust
+{{#include ../ci/concurrency/examples/cs3.rs}}```
+
+### Runtime initialization
+
+A common need in embedded Rust programs is moving, at runtime, a value from
+`main` into an interrupt handler. This can be accomplished at zero cost by
+enforcing sequential access to `static mut` variables.
+
+``` rust
+{{#include ../ci/concurrency/examples/init.rs}}```
+
+In this pattern is important to disable interrupts before yielding control to
+the user program and enforcing that the end user initializes all the
+uninitialized static variables before interrupts are re-enabled. Failure to do
+so would result in interrupt handlers observing uninitialized static variables.
+
+## Redefining `Send` and `Sync`
+
+The core / standard library defines these two marker traits as:
+
+> `Sync`: types for which it is safe to share references between threads.
+>
+> `Send`: types that can be transferred across thread boundaries
+
+Threads are an OS abstraction so they don't exist "out of the box" in bare metal
+context, though they can be implemented on top of interrupts. We'll broaden the
+definition of these two marker traits to include bare metal code:
+
+- `Sync`: types for which it is safe to share references between *execution
+  contexts* that may preempt each other.
+
+- `Send`: types that can be transferred between *execution contexts* that may
+  preempt each other.
```

Comment:
It might make sense to update this wrt. the RFC to restrict Send/Sync reasoning to single cores